iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,192 @@
+{
+  [Logger logMessageWithLevel:level src:src message:message];
+  CHECK_LESS(level, base::g_LogAbortLevel, ("Abort. Log level is too serious", level));
+}
+
+bool AssertMessage(base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:base::LCRITICAL src:src message:message];
+  return true;
+}
+
+// MARK: - Private
+
++ (void)enableFileLogging {
+  Logger * logger = [self logger];
+  NSFileManager * fileManager = [NSFileManager defaultManager];
+
+  // Create a log file if it doesn't exist and setup file handle for writing.
+  NSString * filePath = [[fileManager temporaryDirectory] URLByAppendingPathComponent:kLogFileName].path;
+  if (![fileManager fileExistsAtPath:filePath])
+    [fileManager createFileAtPath:filePath contents:nil attributes:nil];
+
+  NSFileHandle * fileHandle = [NSFileHandle fileHandleForWritingAtPath:filePath];
+  if (fileHandle == nil) {
+    LOG(LERROR, ("Failed to open log file for writing", filePath.UTF8String));
+    [self setFileLoggingEnabled:NO];
+    return;
+  }
+  // Clean up the file if it exceeds the maximum size.
+  if ([fileManager contentsAtPath:filePath].length > kMaxLogFileSize)
+    [fileHandle truncateFileAtOffset:0];
+
+  logger.filePath = filePath;
+  logger.fileHandle = fileHandle;
+  LOG(LINFO, ("File logging is enabled"));
+}
+
++ (void)disableFileLogging {
+  Logger * logger = [self logger];
+
+  [logger.fileHandle closeFile];
+  logger.fileHandle = nil;
+  logger.fileSize = 0;
+  NSError * error;
+  [NSFileManager.defaultManager removeItemAtPath:logger.filePath error:&error];
+  if (error)
+    LOG(LERROR, ("Failed to remove log file", logger.filePath.UTF8String));
+  LOG(LINFO, ("File logging is disabled"));
+}
+
+- (BOOL)isFileLoggingEnabled {
+  return self.fileHandle != nil;
+}
+
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message {
+  LogLevel logLevel;
+  switch (level) {
+    case base::LDEBUG: logLevel = LogLevelDebug; break;
+    case base::LINFO: logLevel = LogLevelInfo; break;
+    case base::LWARNING: logLevel = LogLevelWarning; break;
+    case base::LERROR: logLevel = LogLevelError; break;
+    case base::LCRITICAL: logLevel = LogLevelCritical; break;
+    case base::NUM_LOG_LEVELS:
+      CHECK(false, ("Invalid log level: base::NUM_LOG_LEVELS should not be used in this context."));
+      return;
+  }
+
+  // Build the log message string.
+  auto & logHelper = base::LogHelper::Instance();
+  std::ostringstream output;
+  logHelper.WriteProlog(output, level);
+  logHelper.WriteLog(output, src, message);
+  NSString * logMessage = [NSString stringWithUTF8String:output.str().c_str()];
+
+  Logger * logger = [self logger];
+  // Log the message into the system log.
+  os_log(logger.osLogger, "%{public}s", logMessage.UTF8String);
+
+  dispatch_async([self fileLoggingQueue], ^{
+    // Write the log message into the file.
+    if ([logger isFileLoggingEnabled]) {
```

Comment:
Is this check really needed? Can fileHandle be checked instead? Enabling/disabling logs is a manual, infrequent operation.

Reply:
This check is needed because when the user disables the option, the fileHandle closes the file from writing and is set to nil.
But as we dispatched the file writing to the other serial queue file handle can be nil and we should check it before writing and not before dispatching.